src/main/java/net/trollyloki/advancementbingo/BingoManager.java

```diff
@@ -111,8 +119,87 @@ public void onWin(@NotNull BingoTeam team) {
                 Component.empty(),
                 Title.Times.times(Duration.ofMillis(500), Duration.ofSeconds(3), Duration.ofSeconds(1))
         );
-        for (Player player : Bukkit.getOnlinePlayers())
+        for (Player player : Bukkit.getOnlinePlayers()) {
             player.showTitle(title);
+            Jingles.playGameEnd(plugin, player);
+        }
+    }
+
+    private void onBingoHit(BingoTeam team, NamespacedKey advancement) {
+        if (team.getBoard().isEmpty()) return;
+        @NotNull BingoBoard board = team.getBoard().get();
+
+        int[] advancementLocation = board.findAdvancement(advancement);
+        if (advancementLocation == null) return;
+        int row = advancementLocation[0];
+        int col = advancementLocation[1];
+
+        // Hit for {Team name}
+        // Progress on rows/columns/diagonals
+        // Total bingos scored
+
+        Component hitMessage = hitMessage(team);
+        Component lineMessage = progressMessage(board, row, col);
+        Component completedMessage = Component.text("Completed " + board.getCompletedRows() + "/" + board.getRequiredRows() + " bingos to win")
+                .color(team.getTextColor());
+
+        Bukkit.getScheduler().runTask(plugin, () -> {
```

Comment:
Why is this in a task?

Reply:
It's to make the progress message come out after the vanilla advancement message (because the event fires before the chat message comes out)

Reply:
I see. This is a great example of where a comment would be very good.